app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt

```diff
@@ -42,41 +55,81 @@ class LockboxAutofillService(
+            null, "" -> pslSupport.fromPackageName(packageName)
+            else -> pslSupport.fromWebDomain(webDomain)
         }
 
-        dataStore.list
+        // convert a list of ServerPasswords into a list of (psl+1, ServerPassword)
+        val passwords = dataStore.list
             .take(1)
-            .subscribe { passwords ->
-                val possibleValues = passwords.filter {
-                    it.hostname.contains(webDomain, true)
+            .switchMap { serverPasswords ->
+                val parsedPasswords = serverPasswords
+                    .map { serverPwd ->
+                        pslSupport.fromOrigin(serverPwd.hostname)
+                            .map { FillablePassword(it, serverPwd) }
+                    }
+                val zipper: (Array<Any>) -> List<FillablePassword> = {
+                    it.filter { it is FillablePassword }
+                        .map { it as FillablePassword }
+                }
+
+                when (serverPasswords.size) {
+                    0 -> Observable.just(emptyList<FillablePassword>())
+                    else -> Observable.zipIterable(
+                        parsedPasswords,
+                        zipper,
+                        false,
+                        serverPasswords.size
+                    )
                 }
+            }
+            .take(1)
+
+        val doNext: (Pair<PublicSuffix, List<FillablePassword>>) -> Unit = {
+            val expected = it.first
+            if (expected.isEmpty()) {
+                callback.onFailure("no web domain to match against")
+            } else {
+                val possibleValues = it.second
+                    .filter { fillable ->
+                        fillable.domain.isNotEmpty() && fillable.domain.matches(expected)
+                    }
+                    .map { fillable -> fillable.entry }
                 val response = buildFillResponse(possibleValues, parsedStructure)
                 if (response == null) {
                     callback.onFailure("no logins found for this domain")
                 } else {
+                    log.debug("autofill searching response")
                     callback.onSuccess(response)
                 }
             }
-            .addTo(compositeDisposable)
-    }
+        }
+        val doError: (Throwable) -> Unit = {
+            val msg = "autofill searching unexpectedly failed"
+            log.error(msg, it)
+            callback.onFailure(msg)
+        }
 
-    private fun domainFromPackage(packageName: String): String? {
-        // naively assume that the `y` from `x.y.z`-style package name is the domain
-        // untested as we will change this implementation with issue #375
-        val domainRegex = Regex("^\\w+\\.(\\w+)\\..+")
-        return domainRegex.find(packageName)?.groupValues?.get(1)
+        Observables.combineLatest(expectedDomain, passwords)
+            .subscribe(doNext, doError)
```

Comment:
I personally would include the `doNext` and `doError` blocks inline with my subscription, but I think that's more of a style choice than a hard requirement.